File: backend/app/schemas/urls.py
"""
URL Schemas for API responses.

Defines Pydantic models for URL data from the urls table.
URLs are discovered by Katana, Waymore, GAU, etc. and probed by url-resolver.

Author: Pluckware Development Team
Date: December 2025
"""

from typing import List, Optional, Dict, Any, Union
from pydantic import BaseModel
from datetime import datetime


class URLResponse(BaseModel):
    """Response model for a single URL record."""
    id: str
    asset_id: str
    scan_job_id: Optional[str] = None
    
    # Core URL data
    url: str
    url_hash: str
    domain: str
    path: Optional[str] = None
    query_params: Optional[Dict[str, Any]] = None
    
    # Discovery tracking
    sources: List[str] = []
    first_discovered_by: Optional[str] = None
    first_discovered_at: Optional[datetime] = None
    
    # Resolution metadata
    resolved_at: Optional[datetime] = None
    is_alive: Optional[bool] = None
    status_code: Optional[int] = None
    content_type: Optional[str] = None
    content_length: Optional[int] = None
    response_time_ms: Optional[int] = None
    
    # Enrichment data
    title: Optional[str] = None
    final_url: Optional[str] = None
    redirect_chain: Optional[List[Any]] = None  # Can be status codes (int) or URLs (str)
    webserver: Optional[str] = None
    technologies: Optional[List[str]] = None
    
    # Classification
    has_params: bool = False
    file_extension: Optional[str] = None
    
    # Timestamps
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class URLStatsResponse(BaseModel):
    """Response model for URL statistics."""
    total_urls: int = 0
    alive_urls: int = 0
    dead_urls: int = 0
    pending_urls: int = 0
    urls_with_params: int = 0
    unique_domains: int = 0
    top_sources: List[Dict[str, Any]] = []
    top_status_codes: List[Dict[str, Any]] = []
    top_technologies: List[Dict[str, Any]] = []
    top_file_extensions: List[Dict[str, Any]] = []

    class Config:
        from_attributes = True


class PaginatedURLResponse(BaseModel):
    """Response model for paginated URL results."""
    urls: List[URLResponse]
    total: int
    page: int
    per_page: int
    total_pages: int

    class Config:
        from_attributes = True

File: frontend/src/types/urls.ts
/**
 * URL Resolver Types
 * 
 * Types for the urls table managed by the url-resolver module.
 * URLs are discovered by Katana, Waymore, GAU, etc. and probed/enriched by url-resolver.
 */

export interface URLRecord {
  id: string;
  asset_id: string;
  scan_job_id: string | null;
  
  // Core URL data
  url: string;
  url_hash: string;
  domain: string;
  path: string;
  query_params: Record<string, string> | null;
  
  // Discovery tracking (sources not exposed via API)
  first_discovered_at: string;
  
  // Resolution metadata
  resolved_at: string | null;
  is_alive: boolean | null;
  status_code: number | null;
  content_type: string | null;
  content_length: number | null;
  response_time_ms: number | null;
  
  // Enrichment data
  title: string | null;
  final_url: string | null;
  redirect_chain: string[] | null;
  webserver: string | null;
  technologies: string[] | null;
  
  // Classification
  has_params: boolean;
  file_extension: string | null;
  
  // Timestamps
  created_at: string;
  updated_at: string;
}

export interface URLStats {
  total_urls: number;
  alive_urls: number;
  dead_urls: number;
  pending_urls: number;
  urls_with_params: number;
  unique_domains: number;
  top_status_codes: { status_code: number; count: number }[];
  top_technologies: { name: string; count: number }[];
  top_file_extensions: { extension: string; count: number }[];
}

export interface URLQueryParams {
  limit?: number;
  offset?: number;
  asset_id?: string;
  is_alive?: boolean;
  status_code?: number;
  has_params?: boolean;
  file_extension?: string;
  search?: string;
}

